migration to Nexus 3.71 API

stklcode opened the pull request:
Update the plugin structures and formats to latest Nexus 3.71 changes, i.e. drop OrientDB support, migrate to fluent API.

**TODO:**
- [x] add DAOs to store data with H2/PostgreSQL backends
- [x] add browse module
- [x] migration of recipes and handlers for hosted Composer
- [x] migrate proxy Composer
- [x] migrate group Composer
- [x] update unit tests
- [x] update integration tests
- [x] cleanup
- [x] **feature:** implement _packages/list.json_ for hosted repository

Run-tested all 3 repository types in a local Nexus 3.71.0-06 container (Java 17, H2)

----

This is based on #154. Resolves #157

cattong commented:
good job!

stklcode commented:
ping @bhamail

Could you have a look at this (or name someone who can)?
CI requirements are misconfigured (underscores vs. hyphens in job name) and he CLA bot is not running :shrug:  But the (modified) tests succeed.

I know this is quite a breaking change, but I guess NXRM 3.71 is breaking for almost any plugin. We might want to add a migration utility :thinking: 
Few people have already given feedback to snapshot builds and there is also little discussion about further changes. But first place we'd like to see the plugin working agein (discussion in #157)

bhamail commented on nexus-repository-composer/src/main/java/org/sonatype/nexus/repository/composer/store/ComposerAssetBlobDAO.java:
```diff
@@ -1,6 +1,6 @@
 /*
  * Sonatype Nexus (TM) Open Source Version
- * Copyright (c) 2018-present Sonatype, Inc.
+ * Copyright (c) 2008-present Sonatype, Inc.
```
?

bhamail commented on pom.xml:
```diff
@@ -68,31 +68,35 @@
+        <configuration>
+          <rules>
+            <!--
+            Ensure use of Java 8, as required by NXRM
```
Not certain about the Java rules going forward, but I think maybe Java 8 actually stops being supported after NXRM 3.70. See "Java" https://help.sonatype.com/en/sonatype-nexus-repository-system-requirements.html

This gets messy fast. 

Maybe add a link to the docs in the comment. At some point, we probably want to PREVENT the use of Java 8...

stklcode replied:
This is just the JDK version required to build the plugin, not necessarily the same as the runtime requirement.

I’d be happy to raise this to 11 or 17. We could also clean up quite a few lines then.

bhamail reviewed (approved):
I am NOT a Nexus Repository expert, so please test these changes thoroughly to make sure they work well for you. 

It is really great to see this contribution!!! There are lot's of community repo format plugins that could benefit from this sort of update to the new DB hotness. Thanks for diving into the deep end!

Another thought: It might be worth adding some documentation that states the current version of the Composer plugin  ONLY works with NXRM v3.70+ or whatever. That way, people on older versions hopefully don't upgrade themselves into a bad place by accident. (There are likely to be a lot of upgrade challenges with the DB changes, and anything we can do to minimize that sort of pain is good(tm) ).

stklcode commented on nexus-repository-composer/pom.xml:
```diff
@@ -5,7 +5,8 @@
     <artifactId>composer-parent</artifactId>
-    <version>0.0.30-SNAPSHOT</version>
+    <version>0.1.0-SNAPSHOT</version>
+    <relativePath>../pom.xml</relativePath>
```
Added this one in the early stages when the -it module didn’t even compile.

It’s a multi-module project, so if you want to build only one module Maven will try to look up the parent in local or remote repositories. Makes no difference on release versions, but you typically want to use the latest local changes for snapshots.

Here you cannot run this
```bash
cd nexus-repository-composer
mvn compile
```
without prior “install“ on the parent POM. And with install Macen will pick up the last installed version and no changes made later.

stklcode commented on nexus-repository-composer/src/main/java/org/sonatype/nexus/repository/composer/internal/browse/ComposerBrowseModule.java:
```diff
@@ -1,6 +1,6 @@
 /*
  * Sonatype Nexus (TM) Open Source Version
- * Copyright (c) 2018-present Sonatype, Inc.
+ * Copyright (c) 2008-present Sonatype, Inc.
```
Oops. Copied the base class from another plugin and replaced it later, so most likely copy&paste, definitely not intentional. Will clean this up by tomorrow.

stklcode commented on nexus-repository-composer-it/src/test/java/org/sonatype/nexus/repository/composer/internal/ComposerProxyIT.java:
```diff
@@ -12,32 +12,35 @@
 import static org.hamcrest.Matchers.is;
 import static org.sonatype.nexus.testsuite.testsupport.FormatClientSupport.status;
 
+@Ignore("this test should either be reworked or removed")
```
Think `@todo` is not valid JavaDoc (JSDoc and PHPdoc). Nevertheless `TODO: ...` will do the job for about any IDE and analysis system, so I've added a few lines.

stklcode commented:
> I am NOT a Nexus Repository expert, so please test these changes thoroughly to make sure they work well for you.

Anyway thanks very much for your review.
I’m testing this on my backup instance and few people here also tested my snapshot builds (feedback already included in the code).

> Another thought: It might be worth adding some documentation that states the current version of the Composer plugin ONLY works with NXRM v3.71+ or whatever. [...]

Updated the README, including a notice on NXRM compatibility, added Java 17 and updated some broken links.

3v01ut10n commented:
Thanks for the new version, it allowed me to update my Nexus

stklcode commented:
Hmm, I thought I fixed the integration tests. Unfortunately the tests are passing locally, but not in CI.

```
[ERROR] Errors: 
[ERROR]   ComposerProxyIT.providersURLChangedToNXRM » ClassNotFound org.sonatype.nexus.testsuite.testsupport.RepositoryITSupport not found by PAXEXAM-PROBE-283c035a-7258-4d01-a77b-293598258c42 [108]
[ERROR]   ComposerProxyIT.retrieveListJSONFromProxyWhenRemoteOnline » ClassNotFound org.sonatype.nexus.testsuite.testsupport.RepositoryITSupport not found by PAXEXAM-PROBE-283c035a-7258-4d01-a77b-293598258c42 [108]
[ERROR]   ComposerProxyIT.retrievePackagesJSONFromProxyWhenRemoteOnline » ClassNotFound org.sonatype.nexus.testsuite.testsupport.RepositoryITSupport not found by PAXEXAM-PROBE-283c035a-7258-4d01-a77b-293598258c42 [108]
[ERROR]   ComposerProxyIT.retrieveProviderJSONFromProxyWhenRemoteOnline » ClassNotFound org.sonatype.nexus.testsuite.testsupport.RepositoryITSupport not found by PAXEXAM-PROBE-283c035a-7258-4d01-a77b-293598258c42 [108]
[ERROR]   ComposerProxyIT.retrieveZipballFromProxyWhenRemoteOnline » ClassNotFound org.sonatype.nexus.testsuite.testsupport.RepositoryITSupport not found by PAXEXAM-PROBE-283c035a-7258-4d01-a77b-293598258c42 [108]
[ERROR]   ComposerProxyIT.unresponsiveRemoteProduces404 » ClassNotFound org.sonatype.nexus.testsuite.testsupport.RepositoryITSupport not found by PAXEXAM-PROBE-283c035a-7258-4d01-a77b-293598258c42 [108]
```
